### PR TITLE
Add Canoe Gen2

### PR DESCRIPTION
**Pull request by evyd13**
## Description

Adds the Canoe Gen2. It has per key RGB using the ws2812 protocol.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — APPROVED** [review]
I'm not really a fan of the folder name.  `/percent/canoe/gen2` or rev2 would be better, I think.

File: keyboards/percent/canoegen2/keymaps/via/rules.mk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Build Options
+#   change yes to no to disable
+#
```
Comment (fauxpark) [inline keyboards/percent/canoegen2/keymaps/via/rules.mk:3]:
```suggestion
```

File: keyboards/percent/canoegen2/canoegen2.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+{ \
+    {k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k0a, k0b, k0c, k0d, k0e}, \
+    {k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k1a, k1b, k1c, k1d, k1e}, \
+    {k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k2a, k2b, k2c, k2d, k2e}, \
+    {k30, k31, k32, k33, k34, k35, k36, k37, k38, k39, k3a, k3b, k3c, k3d, k3e}, \
+    {k40, k41, k42, ___, ___, ___, k46, ___, ___, k49, k4a, ___, k4c, k4d, k4e}  \
+}
+
+void matrix_init_user(void);
+void matrix_scan_user(void);
```
Comment (fauxpark) [inline keyboards/percent/canoegen2/canoegen2.h:57]:
```suggestion
```

File: keyboards/percent/canoegen2/canoegen2.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+bool led_update_kb(led_t led_state) {
+    if(led_update_user(led_state)) {
+        writePin(E6, !led_state.caps_lock);
+    }
+
+    return true;
+}
+
+void suspend_power_down_kb(void) {
```
Comment (fauxpark) [inline keyboards/percent/canoegen2/canoegen2.c:33]:
```suggestion
#ifdef RGB_MATRIX_ENABLE
void suspend_power_down_kb(void) {
```

File: keyboards/percent/canoegen2/canoegen2.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    { 231,  40 }  // top 3
+}, {
+    // LED Index to Flag
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 1,
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 1,
+    1, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 1,
+    1, 1, 1, 4, 4, 1, 1, 1, 1, 1,
+    2, 2, 2
+} };
```
Comment (fauxpark) [inline keyboards/percent/canoegen2/canoegen2.c:139]:
```suggestion
} };
#endif
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/percent/canoegen2/canoegen2.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+#include "quantum.h"
+
+// readability
+#define ___ KC_NO
+
+#define LAYOUT_all LAYOUT_65_iso
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/canoegen2.h:24]:
```suggestion
#define LAYOUT_all LAYOUT_65_iso_blocker_split_bs
```

File: keyboards/percent/canoegen2/canoegen2.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#define LAYOUT_65_ansi( \
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/canoegen2.h:26]:
```suggestion
#define LAYOUT_65_ansi_blocker_split_bs( \
```

File: keyboards/percent/canoegen2/canoegen2.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+) \
+{ \
+    {k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k0a, k0b, k0c, k0d, k0e}, \
+    {k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k1a, k1b, k1c, k1d, k1e}, \
+    {k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k2a, k2b, k2c, k2d, k2e}, \
+    {k30, ___, k32, k33, k34, k35, k36, k37, k38, k39, k3a, k3b, k3c, k3d, k3e}, \
+    {k40, k41, k42, ___, ___, ___, k46, ___, ___, k49, k4a, ___, k4c, k4d, k4e}  \
+}
+
+#define LAYOUT_65_iso( \
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/canoegen2.h:41]:
```suggestion
#define LAYOUT_65_iso_blocker_split_bs( \
```

File: keyboards/percent/canoegen2/info.json [hunk 8]
```diff
@@ -0,0 +1,162 @@
+{
+    "keyboard_name": "Canoe Gen2", 
+    "url": "", 
+    "maintainer": "evyd13", 
+    "width": 16, 
+    "height": 5, 
+    "layouts": {
+        "LAYOUT_65_ansi": {
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/info.json:8]:
```suggestion
        "LAYOUT_65_ansi_blocker_split_bs": {
```

File: keyboards/percent/canoegen2/info.json [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                {"x":2.5, "y":4, "w":1.25},
+                {"x":3.75, "y":4, "w":6.25},
+                {"x":10, "y":4, "w":1.25},
+                {"x":11.25, "y":4, "w":1.25},
+                {"x":13, "y":4},
+                {"x":14, "y":4},
+                {"x":15, "y":4}
+            ]
+        },
+        "LAYOUT_65_iso": {
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/info.json:84]:
```suggestion
        "LAYOUT_65_iso_blocker_split_bs": {
```

File: keyboards/percent/canoegen2/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+[0] = LAYOUT_65_ansi(
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/keymaps/default/keymap.c:22]:
```suggestion
[0] = LAYOUT_65_ansi_blocker_split_bs(
```

File: keyboards/percent/canoegen2/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  KC_ESC,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,   KC_9,     KC_0,    KC_MINS, KC_EQL,   KC_BSLS, KC_BSPC, KC_INS,
+  KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,   KC_O,     KC_P,    KC_LBRC, KC_RBRC,  KC_BSLS,          KC_DEL,
+  KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,   KC_L,     KC_SCLN, KC_QUOT,           KC_ENT,           KC_PGUP,
+  KC_LSFT,          KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,   KC_COMM,  KC_DOT,  KC_SLSH, KC_RSFT,  KC_UP,            KC_PGDN,
+  KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, MO(1),             KC_LEFT, KC_DOWN, KC_RGHT),
+
+[1] = LAYOUT_65_ansi(
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/keymaps/default/keymap.c:29]:
```suggestion
[1] = LAYOUT_65_ansi_blocker_split_bs(
```

File: keyboards/percent/canoegen2/rules.mk [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+RGB_MATRIX_ENABLE = WS2812
+
+LAYOUTS = 65_ansi 65_iso
```
Comment (noroadsleft) [inline keyboards/percent/canoegen2/rules.mk:25]:
```suggestion
LAYOUTS = 65_ansi_blocker_split_bs 65_iso_blocker_split_bs
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
> I'm not really a fan of the folder name. `/percent/canoe/gen2` or rev2 would be better, I think.

I agree with @drashna about this. The folder name should definitely be `percent/canoe/gen2` or `percent/canoe/rev2`.

----

## Summary of Requested Changes

- rename `LAYOUT_65_ansi` to `LAYOUT_65_ansi_blocker_split_bs`
- rename `LAYOUT_65_iso` to `LAYOUT_65_iso_blocker_split_bs`
- update Community Layout support per the above (though based on keyboard renders `65_(ansi|iso)` layouts are almost undoubtedly supported)

**Comment by kb-elmo** [comment]
> 
> 
> > I'm not really a fan of the folder name. `/percent/canoe/gen2` or rev2 would be better, I think.
> 
> I agree with @drashna about this. The folder name should definitely be `percent/canoe/gen2` or `percent/canoe/rev2`.

Imho no because that would give the impression that the boards are in some way related to each other design or pcb wise which is entirely not the case.
The Gen 2 is a complete redesign with an entirely different PCB (the gen 1 used a ps2avr PCB).
The only similarity between these boards is that they are called Canoe.

**Comment by fauxpark** [comment]
The same is technically true of the Planck and Preonic...

**Comment by kb-elmo** [comment]
Putting the boards in one folder would be like putting Keycult boards in one folder named TKL with subfolders named No1 and No2 because "technically" they are just another version of the Keycult TKL.

It doesn't make sense because it's not just some revision of the board.
The first board was called Canoe. The second Canoe Gen2. 
So it makes sense to put it in a folder called canoegen2.

**Comment by drashna** [comment]
> The first board was called Canoe. The second Canoe Gen2.

Both boards use the same layout, correct?  If so, then revision works fine here.  
That's fauxparks point with the planck and preonic.  Specifically, there are 4 different planck boards, the Rev1-5, the Rev6, the EZ, and the Light.  They're all "planck" but the matrix and features and controller are all different. Rev1-5 uses atmega32u4, the rev6 and EZ use stm32f303, and the light uses at90usb1286. The EZ and Light use per key RGB, using different ISSI controllers), etc.

So having a different controller, a different matrix and different features really isn't a good reason to have a completely different name and folder.  

And there are plenty of other boards that follow this pattern too (for instance, the helix).  



> So it makes sense to put it in a folder called canoegen2.

IMO, no it doesn't, given the precedent already in the repo. 

However `canoe/gen2` would also be fine.  and looks better than `canoegen2`.

**Comment by Erovia** [comment]
>  So it makes sense to put it in a folder called canoegen2.

Not just looks better, but is what the user's familiar with QMK are expecting.

**Comment by jackhumbert** [comment]
> It doesn't make sense because it's not just some revision of the board.

I think "gen" is functionally the same as "rev" here, so should get the same treatment. 

Usually when introducing another revision, what code was previously just `keyboard/` gets moved to `keyboard/rev1/`, and any shared code can be moved to back to `keyboard/`. This allows folks to easily use the same keymap between both revisions without having to copy it into the new keyboard's folder.

**Comment by kb-elmo** [comment]
> 
> 
> > It doesn't make sense because it's not just some revision of the board.
> 
> I think "gen" is functionally the same as "rev" here, so should get the same treatment.
> 

No it is not. You apparently didn't read what I said.
This is not a "revision" of the board.
It's an entirely new design and should be treated as such.

And like I said the boards do not share anything but a part of the name.

**Comment by kb-elmo** [comment]
> 
> > The first board was called Canoe. The second Canoe Gen2.
> 
> Both boards use the same layout, correct? If so, then revision works fine here.
> 

This doesn't make any sense at all.
Just because a board has the same layout doesn't mean that it's in any way a similar keyboard.
By that logic we could just place all boards by the same designer in folders named after the layouts.

>
> However `canoe/gen2` would also be fine. and looks better than `canoegen2`.
>

The planck is still the same board and all the PCBs are compatible with the Planck case.
This is simply not the case with the Canoe Gen2.

Again: This is an entirely new design and the boards are in no way compatible with each other.
The Gen2 PCB does **not** fit the original Canoe case and vice versa.
Placing the two boards in the same folder would imply that the boards are compatible with each other.

**Comment by fauxpark** [comment]
> it would give the impression that the boards are in some way compatible with each other which is not the case.

The only compatibility that really matters here is the layout support, as we aim to abstract away the hardware-specific stuff for exactly this reason.

> The planck example does not apply here because the planck with all it's PCB revisions is still the same board

As Drashna mentioned above, the different revisions of the Planck use different MCUs (and I believe the rev6 was designed by someone else whereas most/all of the earlier revs were designed by Wilba). So I think it very much applies here.

The very fact that this board is named "Gen2" kind of *implies* that a gen1 also exists. I'd argue that is more confusing to users.

**Comment by kb-elmo** [comment]
> 
> 
> > it would give the impression that the boards are in some way compatible with each other which is not the case.
> 
> The only compatibility that really matters here is the layout support, as we aim to abstract away the hardware-specific stuff for exactly this reason.
> 
> > The planck example does not apply here because the planck with all it's PCB revisions is still the same board
> 
> As Drashna mentioned above, the different revisions of the Planck use different MCUs (and I believe the rev6 was designed by someone else whereas most/all of the earlier revs were designed by Wilba). So I think it very much applies here.
> 
> The very fact that this board is named "Gen2" kind of _implies_ that a gen1 also exists. I'd argue that is more confusing to users.

"Canoe Gen2" is the name of the board. It's not "Canoe" generation 2 or something. Yes technically this could be seen like this but it is meant as the name of the board here.

And again it is not just the MCU or matrix that is different on the new PCB. It does not fit the original Canoe case.

**Comment by fauxpark** [comment]
So why does the website say "revision"?

As for it not fitting the case - AFAICT it doesn't seem like you can buy the PCB on its own, so that doesn't matter anyway.

**Comment by kb-elmo** [comment]
> 
> 
> So why does the website say "revision"?
> 
> As for it not fitting the case - AFAICT it doesn't seem like you can buy the PCB on its own, so that doesn't matter anyway.

You can't right now but there are more PCBs planned. A hotswap PCB should become available later this year for the board as well as a separately available soldered PCB.

**Comment by jackhumbert** [comment]
After talking to kingnestea on the Percent Studio discord, they're ok with `/percent/canoe_gen2` - if you update the keyboard name to that, that'd be great!

**Review by noroadsleft — CHANGES_REQUESTED** [review]
https://github.com/qmk/qmk_firmware/pull/10344#discussion_r493387459 still needs changing; the `via` keymap won't compile without this change (`LAYOUT_65_iso` doesn't exist).

**Review by fauxpark — APPROVED** [review]

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

**Comment by noroadsleft** [comment]
Thanks!

**Comment by evyd13** [comment]
Thank you!

**Pull request merged**